xbmc/games/addons/input/GameClientController.cpp

```diff
@@ -97,42 +97,42 @@ game_controller_layout CGameClientController::TranslateController() const
   if (!m_digitalButtons.empty())
   {
     controllerStruct.digital_buttons = const_cast<char**>(m_digitalButtons.data());
-    controllerStruct.digital_button_count = m_digitalButtons.size();
+    controllerStruct.digital_button_count = static_cast<unsigned int>(m_digitalButtons.size());
   }
   if (!m_analogButtons.empty())
   {
     controllerStruct.analog_buttons = const_cast<char**>(m_analogButtons.data());
-    controllerStruct.analog_button_count = m_analogButtons.size();
+    controllerStruct.analog_button_count = static_cast<unsigned int>(m_analogButtons.size());
   }
   if (!m_analogSticks.empty())
   {
     controllerStruct.analog_sticks = const_cast<char**>(m_analogSticks.data());
-    controllerStruct.analog_stick_count = m_analogSticks.size();
+    controllerStruct.analog_stick_count = static_cast<unsigned int>(m_analogSticks.size());
```

Comment:
It might be easier to change the types in ```game_controller_layout ```. But that's ofc up to you.

Reply:
The way I see it, size_t is for sizes, unsigned int is for counts. This is because sizes can be very large, e.g. the size of a 6GB array doesn't fit in 32 bits. Counts, on the other hand, are usually bounded. In this case, by how many features you can cram onto hardware. The limiting case being keyboards with ~200 keys, and you probably won't ever find more than 2 analog sticks.

Reply:
You should think twice about using `unsigned` for counts, or anywhere really.

http://soundsoftware.ac.uk/c-pitfall-unsigned.html

https://google.github.io/styleguide/cppguide.html#Integer_Types

> Unsigned integers are good for representing bitfields and modular arithmetic. Because of historical accident, the C++ standard also uses unsigned integers to represent the size of containers - many members of the standards body believe this to be a mistake, but it is effectively impossible to fix at this point. The fact that unsigned arithmetic doesn't model the behavior of a simple integer, but is instead defined by the standard to model modular arithmetic (wrapping around on overflow/underflow), means that a significant class of bugs cannot be diagnosed by the compiler. In other cases, the defined behavior impedes optimization.

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-signed

>  Unsigned arithmetic can yield surprising results if you are not expecting it. 

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-nonnegative

> Choosing unsigned implies many changes to the usual behavior of integers, including modulo arithmetic, can suppress warnings related to overflow, and opens the door for errors related to signed/unsigned mixes. Using unsigned doesn’t actually eliminate the possibility of negative values.

Of course there are differing opinions on this, but as you see both ISOCpp and Google recommend to be wary.

Reply:
The standard mentions sizes, not counts. Modulo warning suppression is a valid argument, but like I said, when a count is bounded below any possible modulus an unsigned int can have, this case of warning suppression is actually desired.

Reply:
> Not sure what you mean, do you have an example?

Analog sticks are a good example. The max is two. Even exotic controllers wouldn't have more analog sticks than fingers. so wraparound is not an issue, and any warnings for this would be superfluous.

Reply:
In what case would you get a wraparound warning for the controller count when using `int`?

Reply:
In the cases from the standards you listed. They suggested using `size_t` instead of `unsigned int` could maybe possibly generate warnings. I don't want to introduce the possibility of warnings.

Reply:
> They suggested using size_t instead of unsigned int could maybe possibly generate warnings.

Yes, it may generate warnings (when warranted usually), but this turns the argument that the linked articles are trying to make around. The problem is that `unsigned` *surpresses* warnings that may be relevant.

> If a controller has 5,000,000,000 analog sticks, it's a data validation error, not a programming error.

Agreed, you should properly validate your data - but I think this is besides the point: Using `unsigned` can have surprising results in calculations etc. This is also true if your controller only has 2 sticks.

Reply:
Somehow I get the feeling that this conversation is quite out of sync and we're probably talking about different things by now. Anyway, all I really wanted to say was that `unsigned` has issues in various circumstances - which I did so from my side we're done here :-)

Reply:
👃 I got your nose ;)